game/scenes/ui/time_line.gd

```diff
@@ -11,23 +11,126 @@ var _current_time_seconds: float
 
+func _add_event_dict(event: Dictionary)->bool:
+	if event.has("type") and event["type"] in EventType.values() and \
+	event.has("time") and event.has("duration"):
+		event_schedule.push_back(event)
+		_sort_timeline_events()
+		_add_timeline_marker(event)
+		return true
+	else:
+		print("malformed event dictionary passed to timeline")
+		print(event)
+	return false
+	
+func _add_timeline_marker(event):
+	var new_sprite
+	if event["type"] == 0:#asteroid/meteor
+		new_sprite = $MarkersTemplates/meteor_shower_marker.duplicate()
+	# calculate where to place it, scaled according to game time
+	var total_seconds_in_game = game_time_length_minutes * 60
+	var timeline_length_in_pixels = $HBoxContainer/TextureRect2.rect_size.x
+	var scale = timeline_length_in_pixels / total_seconds_in_game
+	var seconds_to_pixels = event["time"] * scale
+	new_sprite.position.x = seconds_to_pixels + 20#offset
+	new_sprite.show()
+	timeline_markers[new_sprite] = event["time"]
+	$Markers.add_child(new_sprite)
+
+func _add_event(event_type:int, time:int, duration:int)->bool:
+	var event_dict = {}
+	if event_type in EventType.values():
+		event_dict["type"] = event_type
+	else:
+		print("incorrect event type sent to timeline")
+		print(event_type)
+		return false
+	event_dict["time"] = time
+	event_dict["duration"] = duration
+	event_schedule.push_back(event_dict)
+	_sort_timeline_events()
+	_add_timeline_marker(event_dict)
+	return true
+	
+func _get_next_event() ->Dictionary:
+	_sort_timeline_events()
+	var next_event = null
+	for x in event_schedule:
+		if next_event == null:
+			next_event = x
+		else:
+			if x["time"] < next_event["time"]:
+				next_event = x
+	# no need to compare to current time, the schedule should be pruned
+	return next_event
+
+func _sort_timeline_events():
+	event_schedule.sort_custom(CustomSorter, "sort_ascending")
+	
+func _check_event_schedule():
+	var closest_event = _get_next_event()
+	var time_of_event = closest_event["time"]
+	if _current_time_seconds > time_of_event:
+		if closest_event["type"] == 0:
+			emit_signal("AsteroidShowerEvent", closest_event.duplicate() )
+			event_schedule.erase(closest_event)
+			_clean_event_markers(closest_event)
+			#start one event at a time, even if only a few frames apart.
+			# simpler this way
+			return
+	var time_until_next_event = closest_event["time"] - int(_current_time_seconds)
+	if time_until_next_event < 20:
+		var type = ""
+		if closest_event["type"] == 0:
+			type = "Asteroid shower"
+		print("time until next " + type + " is " + str(time_until_next_event) + " seconds away")
+	
+
+func _clean_event_markers(event_dict):
+	var time_to_match = event_dict["time"]
+	for x in timeline_markers.keys():
+		if time_to_match == timeline_markers[x]:
+			timeline_markers.erase(x)
+			x.queue_free()
+			
 
 func _process(delta):
 	_current_time_seconds += delta
 	if _current_time_seconds >= _next_year_seconds:
 		_next_year_seconds += float(seconds_per_year)
+		_check_event_schedule()
```

Comment:
Why do events only happen at year boundaries?  I don't think checking the event would be slow enough to need to limit it this way.  If this is so we can have a count down (lines 113-118) then let's get rid of the count down.  If there's a reason to keep it, let's discuss a better way to do that.